src/replicache.ts

```diff
@@ -572,23 +591,51 @@ export default class Replicache implements ReadTransaction {
    * mutations.
    */
   async push(): Promise<void> {
-    if (this._pushTimerId) {
-      clearTimeout(this._pushTimerId);
-      this._pushTimerId = 0;
-    }
+    this._clearPushTimer();
     await this._wrapInOnlineCheck(() => this._push(MAX_REAUTH_TRIES), 'Push');
   }
 
   private async _push(maxAuthTries: number): Promise<void> {
+    // Is there already a push in flight?
+    if (this._pushCounter > 0) {
+      // Here we create a promise/resolve pair that will resolve when the
+      // current push is done.
+      //
+      // This whole dance is to collapse incoming calls to push during an in
+      // flight push so that they only cause a single actual push RPC. We also want
+      // the call to push to resolve at the right time.
+      //
+      // - push() is called. Lets call this "pushA"
+      // - pushA is being sent over wire
+      // - push() is called again. Lets call this "pushB"
+      // - push() is called again. Lets call this "pushC"
+      // - pushA comes back and is resolved
+      // - puhsB is being sent over wire
+      // - pushB resolve comes back and is resolved
+      // - pushC resolves
+
+      if (!this._pushResolver) {
+        const {promise, resolve} = resolver();
+        this._pushResolver = resolve;
+        this._pushPromise = (async () => {
+          await promise;
+          this._pushResolver = null;
+          await this._push(maxAuthTries);
+        })();
+      }
+      await this._pushPromise;
+      return;
+    }
+
     let pushResponse;
     try {
-      this._changeSyncCounter(1);
+      this._changeSyncCounters(1, 0);
```

Comment:
It seems like `_pushCounter` can only ever be `0` or `1` - is that so? If so why not get rid of it and use the presence of `_pushResolver` for the same purpose? If it is just because you want to decrement during auth we can get rid of that requirement.

Reply:
This was in preparation for allowing > 1 out going pull/push (which seems to work already by just changing a single if statment).

Reply:
but maybe the collapsing of future calls does not work correctly with that... I can probably change it away from a counter even though I find the counter intuitive.

Reply:
I think that allowing more than one outgoing call is going to be a significant project and could make this logic much different. Like for example we are going to want to try and maintain a smooth rate of requests, handle errors gracefully, etc. I don't know how worth it adding complexity now for that future until we have a sketch on how that will work.